README.md
certbot-dns-leaseweb
====================

A certbot plugin that automates the process of completing a dns-01 challenge
using Leaseweb's [Domain API](https://developer.leaseweb.com/api-docs/domains_v2.html).

Install
-------

`certbot-dns-leaseweb` should be installed via pip, in the same python
environment in which `certbot` is installed:

```
pip install certbot-dns-leaseweb
```

Usage
------

Use the following flags to `certbot` to use and control this plugin:

| Option | Description |
|--------|-------------|
| `--authenticator dns-leaseweb` | **required**, enable and use this plugin. |
| `--dns-leaseweb-credentials` | **required**, Leaseweb API credentials INI file. |
| `--dns-leaseweb-propagation-seconds` | optional, seconds to wait for DNS records to propagate. |


Credentials
------------

A suitable API token can be generated from https://secure.leaseweb.com/api-client-management/.

The token should then be stored in an INI file:

```
dns_leaseweb_api_token = notarealtoken
```

**CAUTION**: These API credentials should be carefully protected from exposure
and unauthorised access.

They could be used to complete DNS challenges, allowing potentially unauthorised
parties to obtain or revoke certificates for your domains.

Examples
--------

Obtain a wildcard certificate for `example.com`:

```
certbot certonly \
  --authenticator dns-leaseweb \
  --dns-leaseweb-credentials leaseweb.ini \
  -d '*.example.com'
```

Obtain a wildcard certificate for `example.com`, with extra time for propagation:

```
certbot certonly \
  --authenticator dns-leaseweb \
  --dns-leaseweb-credentials leaseweb.ini \
  --dns-leaseweb-propagation-seconds 600 \
  -d '*.example.com'
```

Docker
------

If you prefer to run certbot with the `dns-leaseweb` plugin in docker, you can
build a suitable image with `make image`.

The image comes with a partially-completed CLI as its entrypoint:

```
/usr/local/bin/certbot \
  --authenticator=dns-leaseweb \
  --dns-leaseweb-credentials=/etc/letsencrypt/credentials/leaseweb.ini
```


You will need to create a suitable credentials file for the plugin at
`/etc/letsencrypt/credentials/leaseweb.ini` for the instructions below to work.


You can then use it as:

```
docker run --rm -ti \
  --volume "/etc/letsencrypt:/etc/letsencrypt" \
  --volume "/var/lib/letsencrypt:/var/lib/letsencrypt" \
  certbot-dns-leaseweb:latest \
    certonly \
      --dns-leaseweb-propagation-seconds 600 \
      -d '*.example.com'
```

Thanks
------

To [Letsencrypt](https://letsencrypt.org/), and the EFF for [certbot](https://certbot.eff.org/).

This plugin is adapted from https://github.com/ctrlaltcoop/certbot-dns-hetzner,
whose README was also the template for this file.

https://github.com/m42e/certbot-dns-ispconfig was also consulted during
development of this plugin.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.3] - 2021-08-14

### Added

* This changelog.

### Modified

* Requesting certificates for arbitrarily deep subdomains is now supported.

## [0.0.2] - 2021-07-18

### Modified

* Fixed typo in setup.cfg bug tracker URL.

## [0.0.1] - 2021-07-18

Initial release for `certbot-dns-leaseweb`.

[0.0.3]: https://gitlab.com/iwaseatenbyagrue/certbot-dns-leaseweb/-/compare/v0.0.2...v0.0.3
[0.0.2]: https://gitlab.com/iwaseatenbyagrue/certbot-dns-leaseweb/-/compare/v0.0.1...v0.0.2
[0.0.1]: https://gitlab.com/iwaseatenbyagrue/certbot-dns-leaseweb/


tests/test_utils.py
"""
Unit tests for utils
"""

import unittest

from certbot_dns_leaseweb import utils


class UtilTest(unittest.TestCase):
    """ Test suite for certbot_dns_leaseweb.utils.
    """

    def test_to_fqdn(self):
        """ feature: return . terminated name when not using shortnames.

        """

        test_cases = [
            {"in": "test", "out": "test"},
            {"in": "test.test", "out": "test.test."},
            {"in": "test.test.", "out": "test.test."},
        ]

        for test_case in test_cases:
            self.assertEqual(utils.to_fqdn(test_case["in"]), test_case["out"])


tests/test_client.py
"""
Unit tests for LeasewebClient
"""

import secrets
import unittest

import requests_mock

from certbot_dns_leaseweb.client import (
    LeasewebClient,
    DomainNotFoundException,
    LEASEWEB_DOMAIN_API_ENDPOINT,
    LEASEWEB_DOMAIN_API_LIST_LIMIT
)


class LeasewebClientTest(unittest.TestCase):
    """ Test suite for certbot_dns_leaseweb.client.LeasewebClient.
    """

    record_domain = "test.test"
    record_name = "test"
    record_content = ["test content"]
    record_ttl = 60

    api_token = "notarealtoken"

    def setUp(self):
        self.client = LeasewebClient(self.api_token)

    def test_add_record(self):
        """ feature: create a DNS record.

        Given a domain name, a record name, and some content
        When I add a new DNS record
        Then a new DNS record should be created
        And it should be of type 'TXT' by default.
        """
        with requests_mock.Mocker() as mock:
            # Mock the response for `LeasewebClient#domains`.
            mock.get(
                (
                    f"{LEASEWEB_DOMAIN_API_ENDPOINT}"
                    f"?offset=0"
                    f"&limit={LEASEWEB_DOMAIN_API_LIST_LIMIT}"
                    f"&type=dns"
                ),
                status_code=200,
                json={
                    "domains": [{"domainName": self.record_domain}],
                    "_metadata": {"totalCount": 1}
                },
            )

            mock.post(
                str(
                    f"{LEASEWEB_DOMAIN_API_ENDPOINT}"
                    f"/{self.record_domain}/resourceRecordSets"
                ),
                status_code=201,
            )
            # Default type and ttl
            self.client.add_record(
                self.record_domain,
                self.record_name,
                self.record_content
            )
            # Explicit type and TTL
            self.client.add_record(
                self.record_domain,
                self.record_name,
                self.record_content,
                "TXT",
                self.record_ttl
            )

    def test_delete_record(self):
        """ feature: delete a DNS record

        Given a domain name and a record name
        When I delete a DNS record
        Then the DNS record should be removed
        And it should be of type 'TXT' by default.
        """
        with requests_mock.Mocker() as mock:
            # Mock the response for `LeasewebClient#domains`.
            mock.get(
                (
                    f"{LEASEWEB_DOMAIN_API_ENDPOINT}"
                    f"?offset=0"
                    f"&limit={LEASEWEB_DOMAIN_API_LIST_LIMIT}"
                    f"&type=dns"
                ),
                status_code=200,
                json={
                    "domains": [{"domainName": self.record_domain}],
                    "_metadata": {"totalCount": 1}
                },
            )

            mock.delete(
                (
                    f"{LEASEWEB_DOMAIN_API_ENDPOINT}/"
                    f"{self.record_domain}/resourceRecordSets/"
                    f"{self.record_name}/TXT"
                ),
                status_code=204,
            )

            # Default type
            self.client.delete_record(
                self.record_domain,
                self.record_name,
            )
            # Explicit type
            self.client.delete_record(
                self.record_domain,
                self.record_name,
                "TXT",
            )

    def test_get_managed_domain_name(self):
        """ feature: Strip subdomains to find owned domain

        Given I have a domain of "example.com"
        When I request a certificate for "subdomain.example.com"
        Then the certbot plugin should add a record to "example.com"
        """

        with requests_mock.Mocker() as mock:
            # Mock the response for `LeasewebClient#domains`.
            mock.get(
                (
                    f"{LEASEWEB_DOMAIN_API_ENDPOINT}"
                    f"?offset=0"
                    f"&limit={LEASEWEB_DOMAIN_API_LIST_LIMIT}"
                    f"&type=dns"
                ),
                status_code=200,
                json={
                    "domains": [{"domainName": self.record_domain}],
                    "_metadata": {"totalCount": 1}
                },
            )

            for subdomain in ["test", "test.test", "test.test.test"]:
                managed_domain = self.client._get_managed_domain_name(
                    f"{subdomain}.{self.record_domain}"
                )

                assert managed_domain == self.record_domain

            with self.assertRaises(DomainNotFoundException):
                self.client._get_managed_domain_name(secrets.token_hex(32))


if __name__ == '__main__':
    unittest.main()


tests/test_plugin.py
"""
Unit tests for LeasewebClient
"""

import unittest

from unittest import mock

from certbot.compat import os
from certbot.plugins import dns_test_common
from certbot.plugins.dns_test_common import DOMAIN
from certbot.tests import util as test_util

from certbot_dns_leaseweb.plugin import (
    LeasewebAuthenticator
)


class LeasewebAuthenticatorTest(
    test_util.TempDirTestCase,
    dns_test_common.BaseAuthenticatorTest,
):
    """ Test suite for `certbot_dns_leaseweb.plugin.LeasewebAuthenticator`.
    """

    def setUp(self):
        super().setUp()

        path = os.path.join(self.tempdir, "file.ini")
        dns_test_common.write(
            {
                "leaseweb_dns_api_token": "notarealtoken",
            },
            path,
        )

        self.config = mock.MagicMock(
            leaseweb_dns_credentials=path,
            # don't wait during tests
            leaseweb_dns_propagation_seconds=0,
        )
        self.auth = LeasewebAuthenticator(self.config, "leaseweb_dns")
        self.mock_client = mock.MagicMock()
        # _get_client | pylint: disable=protected-access
        self.auth._get_client = mock.MagicMock(return_value=self.mock_client)

    @test_util.patch_display_util()
    def test_perform(self, mock_client):
        """ feature: performing a DNS challenge should create a DNS record.

        Given a DNS-01 challenge
        When the LeasewebAuthenticator is used
        Then a new TXT record should be added to the appropriate domain.
        """
        self.auth.perform([self.achall])
        self.mock_client.add_record.assert_called_with(
            DOMAIN, "_acme-challenge." + DOMAIN, mock.ANY
        )

    def test_cleanup(self):
        """ feature: cleaning up a DNS challenge should delete a DNS record.

        Given a DNS-01 challenge has been performed
        When the LeasewebAuthenticator is cleaning up
        Then a TXT record should be deleted from the appropriate domain.
        """

        # _attempt_cleanup | pylint: disable=protected-access
        self.auth._attempt_cleanup = True
        self.auth.cleanup([self.achall])

        expected = [
            mock.call.delete_record(
                DOMAIN, "_acme-challenge." + DOMAIN
            )
        ]
        self.assertEqual(expected, self.mock_client.mock_calls)


if __name__ == '__main__':
    unittest.main()


src/certbot_dns_leaseweb/plugin.py
""" DNS Authenticator plugin for Leaseweb DNS. """

import requests
import zope.interface

from certbot import errors
from certbot import interfaces
from certbot.plugins import dns_common

from certbot_dns_leaseweb.client import (
    LeasewebClient,
    LeasewebException,
)


@zope.interface.implementer(interfaces.IAuthenticator)
@zope.interface.provider(interfaces.IPluginFactory)
class LeasewebAuthenticator(dns_common.DNSAuthenticator):
    """ DNS Authenticator for Leaseweb.

    This Authenticator uses the Leasweb Domains API to complete dns-01
    challenges.
    """

    description = str(
        "Obtain certificates using a DNS TXT record (if using Leaseweb DNS)."
    )

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.credentials = None

    @classmethod
    def add_parser_arguments(cls, add):     # pylint: disable=arguments-differ
        super().add_parser_arguments(add)
        add('credentials', help='Leaseweb credentials INI file.')

    def more_info(self) -> str:
        """ A human-readable string to inform users of what this plugin does.
        """
        return "This plugin completes dns-01 challenges using "\
               "the Leaseweb Domains API to configure DNS TXT records."

    def _setup_credentials(self):
        self.credentials = self._configure_credentials(
            "credentials",
            "Leaseweb credentials INI file",
            {
                "api_token":
                    str(
                        "an API token obtained from "
                        "'https://secure.leaseweb.com/api-client-management/'"
                    ),
            },
        )

    def _perform(self, domain: str, validation_name: str, validation: str):
        try:
            self._get_client().add_record(
                domain,
                validation_name,
                [validation],
            )
        except (
                requests.ConnectionError,
                LeasewebException,
        ) as exception:
            raise errors.PluginError(exception)

    def _cleanup(self, domain: str, validation_name: str, validation: str):
        self._get_client().delete_record(
            domain,
            validation_name,
        )

    def _get_client(self):
        return LeasewebClient(self.credentials.conf("api_token"))


src/certbot_dns_leaseweb/client.py
"""
A helper for interacting with the Leaseweb Domains API (v2).

This client supports the minimal subset of it needed to complete dns-01
challenges.

See https://developer.leaseweb.com/api-docs/domains_v2.html for the full API.
"""

from typing import List
from certbot_dns_leaseweb.utils import to_fqdn


import requests



LEASEWEB_DOMAIN_API_ENDPOINT="https://api.leaseweb.com/hosting/v2/domains"
# https://developer.leaseweb.com/api-docs/domains_v2.html#operation/post/domains/{domainName}/resourceRecordSets
LEASEWEB_VALID_TTLS=[
    60, 300, 1800, 3600, 14400, 28800, 43200, 86400
]
# https://developer.leaseweb.com/api-docs/domains_v2.html#operation/get/domains
LEASEWEB_REQUEST_LIMIT=10


class LeasewebException(Exception):
    """ Base exception for LeasewebClient.
    """


class NotAuthorisedException(LeasewebException):
    """ Authorisation exception, indicating a 401 or 403 status code.
    """

    def __init__(
        self,
        *args
    ):
        super().__init__(
            "Missing or invalid API token",
            *args
        )


class RecordNotFoundException(LeasewebException):
    """ Domain or domain record missing exception, indicating a 404 status code.
    """

    def __init__(
        self,
        domain,
        name,
        *args
    ):
        super().__init__(
            f"No such record for domain {domain}: {name}",
            *args
        )


class ValidationFailureException(LeasewebException):
    """ Validation exception, indicating a 400 status code.

    This is typically the result of invalid/unsuitable record content data.
    """

    def __init__(
        self,
        *args
    ):
        super().__init__(
            "Invalid record.",
            *args
        )

class InvalidTTLException(LeasewebException):
    """ Exception indicating that the requested TTL is not permitted.

    Leasweb's API allows a small number of predefined TTL values, this exception
    indicates the requested TTL is not one of the allowed values.
    """

    def __init__(
        self,
        *args
    ):
        super().__init__(
            f"Valid TTL values are {','.join(LEASEWEB_VALID_TTLS)}",
            *args
        )

class DomainNotFoundException(LeasewebException):
    """ Exception indicating we can not determine the assigned domain.

    If the domain can not be found in the domain list, we do not know
    where to create/delete the records for your domain name.
    """
    def __init__(
        self,
        *args
    ):
        super().__init__(
                "Domain not found",
                *args
        )

class LeasewebClient:
    """ A helper dealing with the parts of the Leaseweb Domains API needed to
    complete dns-01 challenges.
    """


    def __init__(
        self,
        token: str
    ):
        """
        Initialise client by providing a valid API token.

        :param token:   Leaseweb API token, can be generated from
                        https://secure.leaseweb.com/api-client-management/.
        """

        self.token = token
        self.session = requests.Session()
        self.session.headers.update(self.headers)
        self._api_endpoint = LEASEWEB_DOMAIN_API_ENDPOINT


    @property
    def headers(
        self
    ):
        """ Return headers added to all/each API request sent by this client.
        """
        return {
            "Content-Type": "application/json",
            "X-LSW-Auth": self.token,
        }

    def _get_all_zones(self):
        """ Gets all the domains from the api

        :raises .NotAuthorisedException: API token is either invalid, or not
                                         authorised to perform the requested
                                         operation.
        :raises .LeasewebException: Whenever an unexpected error occurs from the api.
        """
        domains = []
        offset = 0

        while True:
            response = self.session.get(
                f"{self._api_endpoint}?limit={LEASEWEB_REQUEST_LIMIT}&offset={offset}"
            )

            if response.status_code == 200:
                domains.extend([e["domainName"] for e in response.json()["domains"]])

                total_count = response.json()["_metadata"]["totalCount"]
                offset += LEASEWEB_REQUEST_LIMIT
                if offset > total_count:
                    break

            elif response.status_code in [401, 403]:
                raise NotAuthorisedException()

            LeasewebClient._report_error(response)

        return domains

    @classmethod
    def _report_error(cls, response):
        """ Switches between leaseweb errors (preferred) and http exceptions.

        If leaseweb gives us an error (in json), use that. Otherwise,
        fall back to the response reason.
        """

        if "error_message" in response.json():
            raise LeasewebException(response.json()["error_message"])

        raise LeasewebException(response.reason)

    @classmethod
    def _domain_sort(cls, domain):
        """ Sort domains. Use this with sorted(<domains>, key=self._domain_sort)
        """
        return list(reversed(domain.split('.')))


    def _find_managed_zone(self, domain):
        """ Find the longest matching domain in the api.

        :param domain: The name of the domain to find managed zone for.

        :raises DomainNotFoundException: The managed domain could not be found.
        """

        domains = self._get_all_zones()

        for zone_name in reversed(sorted(domains, key=LeasewebClient._domain_sort)):
            if domain.endswith(zone_name):
                return zone_name

        raise DomainNotFoundException(domain)


    def delete_record(
        self,
        domain_name: str,
        name: str,
        record_type: str = "TXT"
    ):
        """ Delete a DNS record given its domain, name, and type.

        :param domain_name: The name of the domain to delete the record from.
        :param name: The name of the record to delete.
        :param record_type: The type of record (TXT,A, etc) to delete. Default
                            is "TXT".

        :raises .RecordNotFoundException: The specified domain or record could
                                          not be found.
        :raises .NotAuthorisedException: API token is either invalid, or not
                                         authorised to perform the requested
                                         operation.

        """

        zone_id = self._find_managed_zone(domain_name)

        fqdn = to_fqdn(name)
        response = self.session.delete(
            f"{self._api_endpoint}/{zone_id}/resourceRecordSets/{fqdn}/{record_type}"
        )

        if response.status_code == 204:
            return

        if response.status_code == 404:
            raise RecordNotFoundException(domain=domain_name, name=name)

        if response.status_code in [401,403]:
            raise NotAuthorisedException()

        raise LeasewebException(response.json["error_message"])


    def add_record(
        self,
        domain_name: str,
        name: str,
        content: List[str],
        record_type: str = "TXT",
        ttl: int = 60,
    ):  # pylint: disable=too-many-arguments

        """ Create a DNS record for a domain from name, content, and type data.

        :param domain_name: The name of the domain to add the record to.
        :param name: The name of the record to add.
        :param content: A list of 1 or more strs to populate the record with.
        :param record_type: The type of record (TXT,A, etc) to add. Default
                            is "TXT".
        :param ttl: The TTL of the record in seconds. Default is 60,
                    valid values are 60, 300, 1800, 3600, 14400, 28800, 43200
                    and 86400.
        :raises .InvalidTTLException: Specified TTL is not one of the allowed
                                      values.
        :raises .ValidationFailureException: Indicates the record name or content
                                             may not be valid for its domain or
                                             type.
        :raises .NotAuthorisedException: API token is either invalid, or not
                                         authorised to perform the requested
                                         operation.
        """

        if ttl not in LEASEWEB_VALID_TTLS:
            raise InvalidTTLException()

        zone_id = self._find_managed_zone(domain_name)

        response = self.session.post(
            f"{self._api_endpoint}/{zone_id}/resourceRecordSets",
            json={
                "name": to_fqdn(name),
                "type": record_type,
                "ttl": ttl,
                "content": content
            }
        )

        if response.status_code == 201:
            return

        if response.status_code == 400:
            raise ValidationFailureException(response.reason)

        if response.status_code in [401,403]:
            raise NotAuthorisedException()

        LeasewebClient._report_error(response)
